### PR TITLE
Refactor canvas [SCI-1642]

### PR DESCRIPTION
**Pull request by okriuchykhin**
https://biosistemika.atlassian.net/browse/SCI-1642

### CODE REVIEW

File: app/models/experiment.rb [hunk 1]
```diff
@@ -583,42 +564,39 @@ def update_module_connections(connections)
...
+      raise ArgumentError, 'Cycles exist.'
     end
 
     # First, delete existing connections
     # but keep a copy of previous state
     previous_sources = {}
     previous_sources.default = []
-    my_modules.each do |m|
+
+    my_modules.includes(inputs: { from: [:inputs, outputs: :to] }).each do |m|
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/models/experiment.rb]: It needs all tasks to get 'previous_sources' and later unassign samples from deleted tasks

**Comment by okriuchykhin** [comment]
On medium sized canvas time in ActiveRecord dropped from ~700ms to ~200ms. But I think main problem was in generation of module group name which used recursion.

**Comment by Ducz0r** [comment]
It doesn't affect any of the addons (namely billing), right?

**Comment by okriuchykhin** [comment]
One PR in electronic signatures

**Pull request merged**